services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java

```diff
@@ -2121,17 +2124,70 @@ public void rollForwardToFutureVersion(String clusterName, String storeName, Str
       message.operationType = AdminMessageType.ROLLFORWARD_CURRENT_VERSION.getValue();
       message.payloadUnion = rollForwardCurrentVersion;
 
-      Map<String, String> futureVersions = getFutureVersionsForMultiColos(clusterName, storeName);
-      int futureVersion = 0;
-      for (Map.Entry<String, String> entry: futureVersions.entrySet()) {
-        futureVersion = Integer.parseInt(entry.getValue());
-        if (futureVersion > 0) {
+      // get the future version from the interested regions which will be used to compare after roll forward
+      Map<String, String> futureVersionsBeforeRollForward = getFutureVersionsForMultiColos(clusterName, storeName);
+      int futureVersionBeforeRollForward = 0;
+      for (Map.Entry<String, String> entry: futureVersionsBeforeRollForward.entrySet()) {
+        if (!isRegionPartOfRegionsFilterList(entry.getKey(), regionFilter)) {
+          continue;
+        }
+        futureVersionBeforeRollForward = Integer.parseInt(entry.getValue());
+        if (futureVersionBeforeRollForward > 0) {
           break;
         }
       }
+
+      if (futureVersionBeforeRollForward <= 0) {
+        throw new VeniceException("Roll forward failed without any future version");
+      }
+
+      LOGGER.info(
+          "Sending roll forward command to future version {} for store {} to child controllers",
+          futureVersionBeforeRollForward,
+          storeName);
       sendAdminMessageAndWaitForConsumed(clusterName, storeName, message);
-      LOGGER.info("Truncating topic {} after rollforward", Version.composeKafkaTopic(storeName, futureVersion));
-      truncateKafkaTopic(Version.composeKafkaTopic(storeName, futureVersion));
+      String kafkaTopic = Version.composeKafkaTopic(storeName, futureVersionBeforeRollForward);
+      LOGGER.info(
+          "Truncating topic {} after child controllers consumed the roll forward messages to not block new versions",
+          kafkaTopic);
+      truncateKafkaTopic(kafkaTopic);
+
+      // check whether the roll forward is successful in all regions in regionFilter.
+      // Add retries to let the child controllers finish processing the roll forward command
+      Map<String, Integer> failedRegions = new HashMap<>();
+      for (int retry = 0; retry < NUMBER_OF_RETRY_FOR_ROLL_FORWARD_CHECK; retry++) {
```

Comment:
can we use `executeWithMaxAttempt` retryutils method to execute the retries? I would say total duration should < 1s

Reply:
Thanks for the pointer, used `executeWithMaxAttempt(.., 5, 200)` which will be ~800ms